Source/JavaScriptCore/wasm/WasmConstExprGenerator.cpp

```diff
@@ -0,0 +1,506 @@
+
+    PartialResult WARN_UNUSED_RETURN addRefFunc(uint32_t index, ExpressionType& result)
+    {
+        if (m_mode == Mode::Evaluate) {
+            JSValue wrapper = m_instance->getFunctionWrapper(index);
+            ASSERT(!wrapper.isNull());
+            result = JSValue::encode(wrapper);
+        }
+        return { };
+    }
+
+    ControlData addTopLevel(BlockSignature signature)
+    {
+        return ControlData(signature);
+    }
+
+    PartialResult WARN_UNUSED_RETURN addBlock(BlockSignature, Stack&, ControlType&, Stack&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addLoop(BlockSignature, Stack&, ControlType&, Stack&, uint32_t) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addIf(ExpressionType, BlockSignature, Stack&, ControlData&, Stack&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addElse(ControlData&, Stack&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addElseToUnreachable(ControlData&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addTry(BlockSignature, Stack&, ControlType&, Stack&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addCatch(unsigned, const TypeDefinition&, Stack&, ControlType&, ResultList&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addCatchToUnreachable(unsigned, const TypeDefinition&, ControlType&, ResultList&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addCatchAll(Stack&, ControlType&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addCatchAllToUnreachable(ControlType&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addDelegate(ControlType&, ControlType&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addDelegateToUnreachable(ControlType&, ControlType&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addThrow(unsigned, Vector<ExpressionType>&, Stack&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addRethrow(unsigned, ControlType&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addReturn(const ControlData&, const Stack&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addBranch(ControlData&, ExpressionType, Stack&) CONST_EXPR_STUB
+    PartialResult WARN_UNUSED_RETURN addSwitch(ExpressionType, const Vector<ControlData*>&, ControlData&, Stack&) CONST_EXPR_STUB
+
+    PartialResult WARN_UNUSED_RETURN endBlock(ControlEntry& entry, Stack& expressionStack)
+    {
+        ASSERT(expressionStack.size() == 1);
+        ASSERT_UNUSED(entry, ControlType::isTopLevel(entry.controlData));
+        m_result = expressionStack.first().value();
+        return { };
+    }
+
+    PartialResult WARN_UNUSED_RETURN addEndToUnreachable(ControlEntry&, Stack&, bool = true) CONST_EXPR_STUB
+
+    PartialResult WARN_UNUSED_RETURN endTopLevel(BlockSignature, const Stack&)
+    {
+        // Some opcodes like "nop" are not detectable by an error stub because the context
+        // doesn't get called by the parser. This flag is set by didParseOpcode() to signal
+        // such cases.
```

Comment:
I implemented this `nop` check this way, but we could probably also add an `addNop` method to all the generators too as a callback from the function parser. In most generators it would do nothing. If preferable, I could make that change.